src/main/java/com/mongodb/spark/sql/connector/schema/BsonDocumentToRowConverter.java

```diff
@@ -0,0 +1,501 @@
+      final String fieldName, final DataType dataType, final BsonValue bsonValue) {
+    return Date.from(convertToInstant(fieldName, dataType, bsonValue).truncatedTo(ChronoUnit.DAYS));
+  }
+
+  private Date convertToTimestamp(
+      final String fieldName, final DataType dataType, final BsonValue bsonValue) {
+    return Date.from(convertToInstant(fieldName, dataType, bsonValue));
+  }
+
+  private float convertToFloat(
+      final String fieldName, final DataType dataType, final BsonValue bsonValue) {
+    return (float) convertToBsonNumber(fieldName, dataType, bsonValue).doubleValue();
+  }
+
+  private double convertToDouble(
+      final String fieldName, final DataType dataType, final BsonValue bsonValue) {
+    return convertToBsonNumber(fieldName, dataType, bsonValue).doubleValue();
+  }
+
+  private int convertToInteger(
+      final String fieldName, final DataType dataType, final BsonValue bsonValue) {
+    return convertToBsonNumber(fieldName, dataType, bsonValue).intValue();
+  }
+
+  private short convertToShort(
+      final String fieldName, final DataType dataType, final BsonValue bsonValue) {
+    return (short) convertToBsonNumber(fieldName, dataType, bsonValue).intValue();
+  }
+
+  private long convertToLong(
+      final String fieldName, final DataType dataType, final BsonValue bsonValue) {
+    return convertToBsonNumber(fieldName, dataType, bsonValue).longValue();
+  }
+
+  private BigDecimal convertToDecimal(
+      final String fieldName, final DataType dataType, final BsonValue bsonValue) {
+    BsonNumber bsonNumber = convertToBsonNumber(fieldName, dataType, bsonValue);
+    switch (bsonNumber.getBsonType()) {
+      case DOUBLE:
+        return BigDecimal.valueOf(bsonNumber.doubleValue());
+      case INT32:
+      case INT64:
+        return BigDecimal.valueOf(bsonNumber.longValue());
+      default:
+        return bsonNumber.decimal128Value().bigDecimalValue();
+    }
+  }
+
+  private byte convertToByte(
+      final String fieldName, final DataType dataType, final BsonValue bsonValue) {
+    return (byte) convertToBsonNumber(fieldName, dataType, bsonValue).longValue();
+  }
+
+  private String convertToString(final BsonValue bsonValue) {
+    switch (bsonValue.getBsonType()) {
+      case STRING:
+        return bsonValue.asString().getValue();
+      case SYMBOL:
+        return bsonValue.asSymbol().getSymbol();
+      case DOCUMENT:
+        return bsonValue.asDocument().toJson(jsonWriterSettings);
+      default:
+        String value = new BsonDocument("v", bsonValue).toJson(jsonWriterSettings);
+        // Strip down to just the value
+        value = value.substring(6, value.length() - 1);
+        // Remove unnecessary quotes of BsonValues converted to Strings.
+        // Such as BsonBinary base64 string representations
+        if (value.startsWith("\"") && value.endsWith("\"")) {
+          value = value.substring(1, value.length() - 1);
+        }
+        return value;
```

Comment:
The code under the `default` label is hacky. Is it possible to use the same approach `BsonDocumentCodec` uses?

1. Get a `Codec` from `CodecRegistry` for `bsonValue.getClass`.
2. Encode into a `JsonWriter` via `Codec.encode`.

Reply:
Unfortunately, using:

```
        StringWriter writer = new StringWriter();
        JsonWriter jsonWriter = new JsonWriter(writer, jsonWriterSettings);
        BSON_VALUE_CODEC.encode(jsonWriter, bsonValue, EncoderContext.builder().build());
        return writer.toString();
   ```
   
Doesn't work as the `AbstractBsonWriter` implementation has lots of precondition checks and causes it to fail. For example: `org.bson.BsonInvalidOperationException: An ObjectId value cannot be written to the root level of a BSON document.`

The implementation is too coupled to writing Bson as the usecase. So the hack work around is the simplest way around it at this time.